### ISSUE TITLE
lib.strings.toInt fails with zero padded integers

### ISSUE THREAD

**Issue by jacoblambda** [NixOS/nixpkgs#173228]
### Describe the bug
Whenever the nixpkg core library's string `toInt` conversion function attempts to convert a string of an integer containing zero padding (ex: `00000123`), it results in a parse error.

### Steps To Reproduce
1. Open a REPL instance with `nix repl '<nixpkgs>'`
2. Evaluate the expression `pkgs.lib.strings.toInt "001"`

### Expected behavior
The expression should evaluate to the integer `1`.

### Current behavior
The expression causes the following error:

```console
error: [json.exception.parse_error.101] parse error at line 1, column 2: syntax error while parsing value - unexpected number literal; expected end of input
```

### Notify maintainers

@Profpatsch @eelco @grahamc @pierron (I grabbed these names from [here](https://github.com/NixOS/nixpkgs/blob/master/maintainers/team-list.nix#L424) however I wasn't exactly sure if they are the right people to ping).

### Metadata

```console
 - system: `"x86_64-linux"`
 - host os: `Linux 5.15.32-gentoo-r1, Gentoo, noversion`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.5.1`
 - channels(lambda): `"nixpkgs-22.05pre340506.5c37ad87222"`
 - nixpkgs: `/home/lambda/.nix-defexpr/channels/nixpkgs`
```

**Comment by srpdt** [comment]
This seems to be caused by a change in the implementation of `builtins.fromJSON` in Nix 2.4 (https://github.com/NixOS/nix/pull/3307). The new parser no longer accepts integers with leading zeroes (as per the JSON standard).

In Nix 2.3.11 repl:
```
nix-repl> builtins.fromJSON "0123"
123
```
In Nix 2.5.0 repl:
```
nix-repl> builtins.fromJSON "0123"
error: [json.exception.parse_error.101] parse error at line 1, column 4: syntax error while parsing value - unexpected number literal; expected end of input
```

**Comment by jacoblambda** [comment]
Hmm. In that case I suppose the paths forward would be to either 

A: Add a note to the docs that leading zeros are not supported and possibly add a `lib.strings.stripLeadingZeros` function.

or

B: Add a `match "0*([[:digit:]]+)"` to `toInt` to filter out leading zeros.

I suspect since it's such an uncommon edge case, option A would be preferable.

**Comment by srpdt** [comment]
> A: Add a note to the docs that leading zeros are not supported and possibly add a `lib.strings.stripLeadingZeros` function.
> 
> or
> 
> B: Add a `match "0*([[:digit:]]+)"` or `match "0*(.+)"` to `toInt` to filter out leading zeros.

Nix itself accepts integers with leading zeroes so it would probably be more intuitive for `toInt` to match that behaviour. Option B also has the benefit of behaving consistently across Nix versions.

**Comment by jacoblambda** [comment]
Resolved with #173949
